### PR TITLE
User type within inserter text

### PR DESCRIPTION
**Pull request by jafeltra**
_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._

I worked with @mgramigna on this.

This pull requests addresses Jira 1555 and supports editing inserter structured phrases. You can now add click or arrow key inside an inserter structured phrase and add new text inside of it. This text will show up black instead of grey since it is not part of the record. Context is only set for the last chuck of original shortcut text. Contexts are displayed correctly in the context tray and breadcrumbs when typing. The new text and contexts remain correct when reopening the note.

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [x] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [x] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [ ] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [x] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [ ] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [ ] Document any new APIs/extension points
- [ ] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- [ ] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by laclark22** [comment]
[LC1] if you go create a new note with the follow up template and then click at the beginning of a line of structured data in oncology history and press enter, when you try and start typing in the blank line above it takes the text down to the line below.

**Comment by mgramigna** [comment]
I could only get LC1 to happen when navigating to that blank line with the arrow keys. If I click on the blank line it works fine.

What this tells me is that the selection object must be handled differently when using the arrow key in the structured field in this case versus clicking on the line. @jafeltra any thoughts?

**Comment by mgramigna** [comment]
[MG1] Copy and pasting structured fields no longer pastes them as structured fields

File: test/backend/views/FullApp.test.js [hunk 1]
```diff
@@ -420,11 +420,11 @@ describe('6 FluxNotesEditor', function() {
 
         // Check structured phrases
         const structuredField = wrapper.find('.structured-field-inserter');
-        expect(structuredField.at(0).text()).to.equal(`Test Name `);
-        expect(structuredField.at(1).text()).to.equal(`49 `);
-        expect(structuredField.at(2).text()).to.equal(`Female `);
+        expect(structuredField.at(0).text().replace(/[^\x20-\x7E]+/g, "")).to.equal(`Test Name`);
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply test/backend/views/FullApp.test.js]: Resolved

File: src/notes/StructuredFieldPlugin.jsx [hunk 2]
```diff
@@ -59,31 +161,31 @@ function StructuredFieldPlugin(opts) {
             structured_field: props => {
                 let shortcut = props.node.get('data').get('shortcut');
                 if (shortcut instanceof InsertValue) {
-                    return <span contentEditable={false} className='structured-field-inserter' {...props.attributes}>{shortcut.getText()}{props.children}</span>;
+                    return <span className='structured-field-inserter' {...props.attributes}>{props.children}&#8203;</span>;
```
[the remark this answers is not in the document]
Comment (mgramigna) [reply src/notes/StructuredFieldPlugin.jsx]: Resolved

**Comment by Dtphelan1** [comment]
[DP1] Spaces are added when I didn't used them to pad out newly added text from fragments of the structured phrases. To replicate: 
1. Insert `@age`
2. Arrow cursor in-between age
3. Insert any letter (e.g. 'a'), leaving you with '5a7')
4. Close note
5. Open note, revealing '5a 7'

**Comment by Dtphelan1** [comment]
[DP2] When my cursor is surrounded by structuredData, special buttons like "End" and "Home" insert the literal strings like 'End' instead of moving my cursor. One way of replicating: 
1.Insert @age
2.Arrow cursor in-between age
3. Hit 'End', resulting in '5End7'

I was also able reproduce this when I was at the beginning of everyline but the first one in @allmedications, leading me to think that the cursor has to be flanked on the left and right by structured data in order for this to happen. Similarly, after the first insertion is doesn't happen again until flanked on both sides by StructuredPhrases.

**Comment by Dtphelan1** [comment]
[DP3] There are corner cases where up-arrowing doesn't move the cursor how I expect. One way of replicating: 
1. Insert @allmedications
2. Move cursor to the end of the structured phrase
3. Click the up-arrow, moving the cursor to the beginning of the line instead of to the line above. 
4. Click the up-arrow again, and nothing happens....

**Comment by Dtphelan1** [comment]
[DP4] Inserting a shortcut from the TDP in-between an existing shortcut formats the newly inserted data with the typical inserter-shortcut styling.

To reproduce: 
1. Insert '@condition' 
2. Choose 'Gastrointestinal...'
3. Move cursor into the middle of 'stromal'
4. Insert "Complete Response" 
Results in: 

image

**Comment by Dtphelan1** [comment]
[DP5/6] I'll start with a snapshot of the end result, but there's two bugs here at least (possibly with the same root cause)

image

Steps to reproduce: 
1. Insert 'Gastro...'
2. Moving your cursor from the beginning of the word toward the end, sprinkle in spaces (e.g. space between 'Gastro' and the rest, then space between ' intest' and the rest, etc...)
3. Close the note
4. Open the note
Observation 1: just has the word 'or'
image

Continuing...
5. Close the note (yes, again) 
6. Open the note
Observation 2: The context now has multiple sections, each devoted to a slice of the original context term

**Comment by mgramigna** [comment]
"We can probably point this down to a 2" -Matt and Julia, 11 December 2018

**Comment by Dtphelan1** [comment]
[DP7] There's a nontrivial problem of how we deal with data that depends on the context of the condition it is splitting. Said another way: What do we do when I insert @ONCOHIST inbetween 'Gastro Int...'. To reproduce: 
1. Insert 'Gastro...'
2. Mouse-click your cursor in the middle somewhere
3. Use the right-hand panel to insert "@oncohist"
4. Move the cursor after insertion.

Right now the following error occurs 
image

**Comment by gregquinn2001** [comment]
We discussed DP7. Given that the context is on the last part of the phrase, the condition shouldn't be in context in the middle. We will need to look into that! It shouldn't allow insertion of @ONCOHIST because there's not conditon context!

**Comment by mgramigna** [comment]
Methinks we addressed DP2

**Comment by mgramigna** [comment]
Some thoughts on the others:

DP1: we were like half aware of this already. Will follow up.

DP2: Resolved.

DP3: We were also aware of this one. We think it has to do with how there are newlines in the shortcut text, but it is represented as a single Slate inline, whereas normally things with newlines span across multiple blocks. This never behaved how we expected it to regardless of any typing in the SF.

DP4: What should the expected behavior be here? This case arises when a value is inserted from the TDP or the context tray in the middle of a word. Should it be plaintext? I feel like it's a pretty far-off use case so we're leaning JIra.

DP5/6: Can't replicate 5. Will follow up with Dylan tomorrow hopefully. 6 we haven't looked into yet but will follow up.

DP7: This is happening because the shortcut itself defines the context, not the cursor position within the shortcut. Do we want to only have context when the cursor is at or after the shortcut text, i.e. disabling context in the middle of a structured field (sounds rough)? @gregquinn2001 

I've been working on a fix for MG1 that behaves pretty strangely in some cases, but it's similar behavior to the copy and paste bugs that exist in the code that's on fluxnotes.org right now anyway. Will probably have to sit down and talk about this because it is not not not fun.

LC1 is still a pretty glaring issue I think, especially cause it's basically the most likely use case to come up in a pilot or something. I believe this is related to what I mentioned for DP3. The newlines in the oncohist shortcut behave strangely with selection and what not since it's all part of one span but has newlines in it.

Yay.

Goodnight everyone.

Love,
Matt

**Comment by gregquinn2001** [comment]
Thanks for the update on all the issues! Let's discuss all of them tomorrow. I hope to be in the office. Maybe after standup we can discuss until my meeting at 10 at least? After my 10 is over though I have nothing else so we can discuss further whenever you guys have time.

Love,
Greg (wow this team is close!)

**Comment by Dtphelan1** [comment]
DP5 is no longer happening on my end after pulling to most recent! So yay!!! :) 

Love, 
Dylan Phelan

**Comment by mgramigna** [comment]
After talking with Greg we decided that my fix for MG1 is in a fine place. Will create two resulting Jiras. Now when you copy only part of a structured field it will paste as plaintext, but when you copy the whole structured field it will paste as a structured field.

**Comment by jafeltra** [comment]
DP7 should be resolved.

**Comment by jafeltra** [comment]
DP6 is resolved!

**Comment by Dtphelan1** [comment]
Gonna look into DP1 today

**Comment by jafeltra** [comment]
LC1 should be fixed.

We fixed it by removing the zero-width space that was included in the rendering of the inserter shortcuts in Chrome. Removing the space fixes the LC1 bug. However, the space is needed in Safari in order for the editor to not crash after inserting a shortcut, which was why it was originally added. Keeping the space when in Safari means that the LC1 bug is still present in Safari, but since note authoring is not a use case for Safari right now and the editor does not crash when typing normally, we decided this fix was sufficient for now. I can make a separate Jira to fix the LC1 bug in Safari.

**Comment by mgramigna** [comment]
I pushed up @Dtphelan1's fix for DP1

**Comment by laclark22** [comment]
[LC2] If you select a follow up template and click into a middle of a line of text in a section and press enter you will get an error if you try and delete all of the text on the new line you have created you get an error.

**Comment by mgramigna** [comment]
[MG2] Very specific, but hitting enter before the last character of a shortcut with context, then arrow keying to the right once, doesn't properly update context and throws a console error

**Comment by mgramigna** [comment]
LC2 isn't fixed. ignore me

**Comment by jafeltra** [comment]
[JA2] We crossed out the new items in the pre-release script since this is not merged. Before merging, update the pre-release script again.

**Comment by mgramigna** [comment]
@laclark22 Made a Jira!

File: src/notes/FluxNotesEditor.jsx [hunk 3]
```diff
@@ -1316,10 +1365,10 @@ class FluxNotesEditor extends React.Component {
                 }
                 remainder = remainder.substring(start + trigger.trigger.length);
 
-                // FIXME: Temporary work around that adds spaces when needed to @-phrases inserted via mic
-                if (start !== 0 && trigger.trigger.startsWith('@') && !before.endsWith(' ')) {
-                    transform = this.insertPlainText(transform, ' ');
-                }
+                // // FIXME: Temporary work around that adds spaces when needed to @-phrases inserted via mic
```
Comment (Dtphelan1) [inline src/notes/FluxNotesEditor.jsx]: Maybe remove this?

Comment (mgramigna) [reply src/notes/FluxNotesEditor.jsx]: Spoke to Dylan about this. We want to leave it in since this was originally in here for dictation. Might need to revisit it in the future so leaving in the commented out code

**Comment by Dtphelan1** [comment]
[DP-X] Repeatedly deleting over multiple blocks throws an error: `Could not find a node with key "n"`

To repeat: 
- Insert @ONCOHIST
- Delete backwards from "White blood result" to "HemoGlobin Lab Result"
- Delete backwards from cursor to "Dec 2018".

Error logs in console.

**Comment by karenzheung** [comment]
Aside from the already commented issues I haven't found anything new. Good job

**Comment by mgramigna** [comment]
this could be it

**Comment by Dtphelan1** [comment]
this could be it   
  
  
  
    
  

  
  
>
